### PR TITLE
Add gym-details sidebar

### PR DESCRIPTION
**Pull request by wobblybobz**
## Description

Add a "View Details" link to the gym popup that opens a sidebar (same as stats sidebar) to display detailed information about the Pokemon in the gym.
I have taken maddhatter's PR1316 and have rebased it, added requested fixes for the original #1316, and have made some font-size and padding adjustments for mobile devices
## Motivation and Context

The information being displayed in the sidebar has been captured since #666, but was not displayed on the front-end.
## How Has This Been Tested?

Linux, Windows 10, mariadb, sqlite and view on mobile chrome and ipad safari
## Screenshots (if appropriate):

1 Free Slot (Red)
Extended Info (blue)
Level 10 (Yellow)
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pogom/app.py [hunk 1]
```diff
@@ -35,6 +35,7 @@ def __init__(self, import_name, **kwargs):
         self.route("/stats", methods=['GET'])(self.get_stats)
         self.route("/status", methods=['GET'])(self.get_status)
         self.route("/status", methods=['POST'])(self.post_status)
+        self.route("/gym_data", methods=['GET'])(self.get_gymdata)
```
Comment (Rikilsta) [inline pogom/app.py]: image

**Comment by wobblybobz** [comment]
I have checked this again and it is working for me, I have just rebased it to the current develop branch. Did you try to merge this with another branch? 
Is the section of code "def get_gymdata(self):" at line 263 still declared?

**Comment by Rikilsta** [comment]
@wobblybobz well i manually editted the files and i double checked that the lines where there. and it was merged with a different branch so i guess will try again with yours.

**Comment by hd1080** [comment]
You should update the screenshots because they still show some problems but you already fixed them in the code.

**Comment by wobblybobz** [comment]
Good idea, Thanks Done...

**Comment by neskk** [comment]
Tested this and seems to be clear of errors, great job!
I would like to suggest a feature: instead of closing the sidebar once you click the map or close the gym tooltip, it would be good if the sidebar had  its own close button and change the behavior to let it stay open even if you close the gym tooltip/click on the map.

**Comment by mew905** [comment]
one complaint: On the moveset of the pokemon, if it happens to be the "Dark" type, it's printed in white text with no background, making it incredibly difficult to read

**Comment by jnovack** [comment]
I have three major design issues with this (I had them on the last PR, never changed.  These gripes do not reflect your choices, you can fix them.):

1) Popup OR Sidebar.  I don't want to have to click a gym, THEN click a link to get me to the Sidebar.  Give me an front-end option to ignore the popup and ONLY show the sidebar.

2) No Scrolling!! If your default layout causes you to scroll on a laptop screen, it's not done right.  There's SO MUCH WASTED SPACE with those padding and borders!!  

3) Show the extra information by default. Every time you want to see the details on the 10th person in a gym (probably the most important one), you have to (1) click the gym, (2) click the View Sidebar, (3) scroll, (4) click the arrow on the 10th person.  *4 STEPS EVERY TIME!*

The percentage is dumb. You are ALREADY showing the IVs; you are already showing something that is more accurate and more valuable.  Wasted space.

**Enough with the clicking!**  Information should be on screen without having to hunt for it.  Hard to reach information is almost as bad as no information.

**Comment by mew905** [comment]
I generally agree with the above, I would love to just click and have the sidebar pop up (but would nullify the empty space for Gyms PR, but I'm sure that wouldn't be difficult to implement), same with the spawn point history sidebar, but I have to disagree on your fourth point: percentage (it wasn't numbered):

I like seeing the Percentage, while it might be quick for some to calculate in their head if 10/14/12 is better than 7/15/14 or not, the percentage gives a good gauge as to how "perfect" a pokemon is. Otherwise, yes, I agree. I've accidentally clicked off the sidebar a few times and had to click-click-click to get back to where I was and found it mildly frustrating. It's just nitpicking IMO, but in a world where we demand instant gratification, that slight nuisance could mean frustrated users.

Personally, I'd like to be able to scroll the map around without the sidebar closing, but with one-click-sidebar opening, it's not a big deal at all.

**Comment by wobblybobz** [comment]
@jnovack Hope your happy with these modifications

1) Hovering over a marker now will show the popup info window, When clicking on it now it will only show the sidebar and the info window will disappear when you move your mouse off the marker

2 The sidebar will not close if you accidentally click on the map.

3) I have reduced some of the padding, please note not every laptop screen is the same resolution so smaller screens will still need to scroll and this is unavoidable. On my laptop where the resolution is 1366x768 i still need to scroll to see the 1 bottom row as per the screenshot updated above. On my other laptop where the resolution is at 1920x1080 I don't need to scroll at all.

4) I will see what i can do about fitting the extra data into the the display, however this may mean you have to do some scrolling or have reduced view of the map. Part of me feels doing this however will make the information too cluttered and hard to read. If you want to mockup your thoughts on what each of the rows could look like we could try that otherwise for now your just going to need to keep clicking.

**Comment by slideup-benni** [comment]
@wobblybobz with your last changes it is impossible to click on the "Get Directions" because when you want to click on it, it disappears.

Maybe you can add the link to the sidebar? And when you do this, it would be very great, when you add last_modified in the sidebar aswell :)

**Comment by hd1080** [comment]
IMO the sidebar sould not open on a mobile device when I just click on the gym icon on the map.

**Comment by jnovack** [comment]
> @wobblybobz said: @jnovack Hope your happy with these modifications

Sorry. I deal with poorly designed software all day. I may have taken my frustrations out on you. My apologies. :/

> On my laptop where the resolution is 1366x768 i still need to scroll to see the 1 bottom row

Which 13" monitor from 1998 are you using?? :) I know some tablets use that resolution.

**Comment by Rook25** [comment]
Should this pr work with the latest version of the map? I haven't been able to get it to show the sidebar anymore.

**Comment by wobblybobz** [comment]
@Rook25 I have tested this, and it is working. There was something that was changed a short time ago, There is now a slide switch under the Show Gyms, to Use Gym Sidebar, as there were some requests for it as some people preferred the popup bubbles over the sidebar.

**Comment by Wilddormouse** [comment]
@wobblybobz 
Great addition, monitoring gyms is now easily been done. 

Is it possible to view the date "Gym last modified" in the sidebar? The date is collected, but it's need a implementation in the frondend. Perhaps you could indicate wich lines needs to be modified.

**Comment by Rook25** [comment]
@wobblybobz Ah, I didn't realise there was that slide switch change. Thanks a lot.

**Comment by irus99** [comment]
Is this working with the latest version?

**Comment by popcornmix** [comment]
> Is this working with the latest version?

Yes.

**Comment by irus99** [comment]
I get the basic gym info with -gi but I can't see detailed info including IVs is there something else I need to enable?

**Comment by popcornmix** [comment]
Have you turned on "use gym sidebar" in marker settings?

**Comment by irus99** [comment]
I don't have that option in marker settings :/ I just have the option for "gyms" which is enabled.

**Comment by irus99** [comment]
do I need to install this particular branch or the latest dev branch has this commit?

**Comment by popcornmix** [comment]
It's an unmerged pull request, so you will have to apply the patch the current tree to get the new behaviour.
It may be best to wait until it is merged.

**Comment by ark71** [comment]
@irus99 Did you run npm install after you did the pull?

The full process is:
git fetch origin pull/1422/head:1422-GymInfo
git checkout 1422-GymInfo
npm install

**Comment by elgatodiablo** [comment]
i havn't been able to get this to show either

**Comment by irus99** [comment]
yeah I guess I'll just wait until it gets merged. Thanks for the help though.

**Comment by k4n30** [comment]
@wobblybobz Any chance of being updated?

**Comment by jtp10181** [comment]
@k4n30 the merge resolution was an easy fix, if there is an imminent chance of this getting merged I could make a new PR (after I figured out how), but otherwise it pointless because things keep changing and stuff like this aren't getting merged right now.

**Comment by Garalor** [comment]
is it possible to add an option for filtering for specific player in gyms?

**Comment by k4n30** [comment]
@jtp10181 - Yeah I've merged it locally, but seeing as this is already a recreated pull request, I wanted to avoid recreating it again for the third time, hence why I was trying to find out if @wobblybobz is still supporting it first

@Garalor - That's probably better suited to its own issue

**Comment by wobblybobz** [comment]
I'm still here, sorry I have been a bit busy lately will try to spend some time today to get this merge conflict resolved.

**Comment by wobblybobz** [comment]
Rebase Done, Merge Conflicts Resolved

**Comment by TackyBrad** [comment]
Would absolutely love to see this merged.

**Comment by aguita** [comment]
This would be great to have merged!! Such a great PR

**Comment by FrostTheFox** [comment]
Merging locally to check.

**Comment by FrostTheFox** [comment]
Can't seem to get to show the "view details" even after rebuilding and remerging multiple times...

**Comment by FrostTheFox** [comment]
Nevermind, i'm an idiot :P

**Comment by FrostTheFox** [comment]
My apologizes for the delay in PRs, a lot of the dev team has been away, and when they're around they're not testing PRs, so basically I'm the sole tester at the moment. I'm trying to get some of the devs to help out, but if not I'll have to get some more helping hands.

**Comment by Imoq** [comment]
I am getting this error message now and then. It seems harmless but annoying. It's appearing on the web servers (not on any workers, of course). I suspect it has to do something about me using locale: fr

```
2016-12-12 22:56:53,200 [  Thread-2549918][           app][   ERROR] Exception on /gym_data [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1988, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1641, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1544, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1639, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1625, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/root/PokemonGo-Map/pogom/app.py", line 367, in get_gymdata
    gym = Gym.get_gym(gym_id)
  File "/root/PokemonGo-Map/pogom/models.py", line 625, in get_gym
    p['move_1_name'] = get_move_name(p['move_1'])
  File "/root/PokemonGo-Map/pogom/utils.py", line 487, in get_move_name
    return i8ln(get_moves_data(move_id)['name'])
  File "/root/PokemonGo-Map/pogom/utils.py", line 483, in get_moves_data
    return get_moves_data.moves[str(move_id)]
KeyError: '242'
```

Can open an issue if required, but wanted to report it here in case anybody knew of a quick solution.

Thanks!

**Pull request merged**